docs/custom_promise_types/groups.py

```diff
@@ -0,0 +1,335 @@
+    def __init__(self, message):
+        self.message = message
+
+
+class Group:
+    def __init__(self, name, gid, members):
+        self.name = name
+        self.gid = gid
+        self.members = members
+
+    
+    @staticmethod
+    def lookup(group):
+        command = ["grep", f"^{group}", "/etc/group"]
+        process = Popen(command, stdout=PIPE, stderr=PIPE)
+        stdout, stderr = process.communicate()
+
+        if (process.returncode == 2):
+            msg = stderr.decode("utf-8").strip().split('\n')[0]
+            raise GroupException(msg)
+
+        entry = stdout.decode("utf-8").strip()
+        if entry:
+            entry = entry.split(':')
+            name = entry[0]
+            gid = entry[2]
+            members = entry[3].split(',') if entry[3] else []
+            return Group(name, gid, members)
+        return None
+
+
+    @staticmethod
+    def create(group):
+        command = ["groupadd", group]
+        process = Popen(command, stdout=PIPE, stderr=PIPE)
+        stdout, stderr = process.communicate()
+
+        if (process.returncode != 0):
+            msg = stderr.decode("utf-8").strip().split('\n')[0]
+            raise GroupException(msg)
+
+
+    @staticmethod
+    def delete(group):
+        command = ["groupdel", group]
+        process = Popen(command, stdout=PIPE, stderr=PIPE)
+        stdout, stderr = process.communicate()
+
+        if process.returncode != 0:
+            msg = stderr.decode("utf-8").strip().split('\n')[0]
+            raise GroupException(msg)
+
+
+    @staticmethod
+    def gid_set(group, gid):
+        command = ["groupmod", "--gid", gid, group]
+        process = Popen(command, stdout=PIPE, stderr=PIPE)
+        stdout, stderr = process.communicate()
+
+        if process.returncode != 0:
+            msg = stderr.decode("utf-8").strip().split('\n')[0]
+            raise GroupException(msg)
+
+
+    @staticmethod
+    def member_add(group, user):
+        command = ["gpasswd", "--add", user, group]
+        process = Popen(command, stdout=PIPE, stderr=PIPE)
+        stdout, stderr = process.communicate()
+
+        if process.returncode != 0:
+            msg = stderr.decode("utf-8").strip().split('\n')[0]
+            raise GroupException(msg)
+
+
+    @staticmethod
+    def member_remove(group, user):
+        command = ["gpasswd", "--delete", user, group]
+        process = Popen(command, stdout=PIPE, stderr=PIPE)
+        stdout, stderr = process.communicate()
+
+        if process.returncode != 0:
+            msg = stderr.decode("utf-8").strip().split('\n')[0]
+            raise GroupException(msg)
+
+
+    @staticmethod
+    def members_set(group, users):
+        command = ["gpasswd", "--members", ",".join(users), group]
+        process = Popen(command, stdout=PIPE, stderr=PIPE)
+        stdout, stderr = process.communicate()
+
+        if process.returncode != 0:
+            msg = stderr.decode("utf-8").strip().split()
+            raise GroupException(msg)
```

Comment:
Please make these actual methods of `Group` objects -- `lookup()` and `create()` might be static methods returning new instances, or move these into a separate module as functions.

Reply:
I'm not quite sure if the changes I've now made, corresponds to what you meant. But at least it's more OOP now, i think.

Reply:
It's exactly what I mean. Very nice!

Reply:
:clap: